src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,822 @@
+    public:
+    iostream_error(const char *msg) : runtime_error(msg)
+    {
+    }
+};
+// on posixio_errno
+class posixio_error : public std::system_error
+{
+    public:
+    posixio_error(const int posix_errno) : system_error(posix_errno, std::system_category())
+    {
+    }
+};
+
+using streambuf = cpbuffer<IOSTREAM_BUFSIZE>;
+
+// fileinout_base is a base for fileout/filein which defines some general things.
+// Public functions are available in both fileout/filein and they does not get defined
+// at those classes.
+class fileinout_base
+{
+    protected:
+    int fd = -1;
+    std::unique_ptr<streambuf> buf;
+
+    public:
+    // Get raw pointer of current buffer
+    // NOTE: The buffer may be null if allocation failed (buffer_failbit will be set in this case)
+    streambuf *get_buf() noexcept;
+
+    // Is current stream's file descriptor open?
+    bool is_open() noexcept;
+
+    // Set file descriptor
+    // NOTE: Setting new fd will REPLACE current fd in the stream and It doesn't
+    // flush data to current fd before replacing it. Also It may result in leaked
+    // file descriptor (because stream use its fd to know what fd should be closed).
+    void set_fd(const int newfd) noexcept;
+
+    // Get current file descriptor
+    int get_fd() noexcept;
+
+    // Close current file descriptor if it is opened
+    bool close() noexcept;
+
+    // Return true if file is opened
+    operator bool() noexcept;
+
+    // Return true if file is closed
+    bool operator!() noexcept;
+};
+
+/*
+ * fileout
+ * =-=-=-=
+ *
+ * This class provides file output functionality. It holds a std::unique_ptr
+ * of streambuf (a fixed-size circular buffer which that's size can be changed at
+ * compile-time) and controls its file descriptor and handle it gracefully when needed.
+ *
+ * Constructors
+ * ------------
+ *
+ * The fileout class has four constructors:
+ *
+ *      // Simple construction
+ *      dio::fileout output_a;
+ *      if (output_a.buffer_failure()) return -1; // Failure in buffer allocation
+ *
+ *      // Construction with moving a pre-allocated unique_ptr of the buffer
+ *      try {
+ *          std::unique_ptr<streambuf> my_buffer(new streambuf);
+ *      }
+ *      catch(const std::bad_alloc &) {
+ *          print_err("Cannot allocate memory for the output buffer!\n");
+ *          return -1;
+ *      }
+ *      dio::fileout output_b(std::move(my_buffer)); // Takes ownership of my_buffer
+ *
+ *      // Construction with a file descriptor number
+ *      dio::fileout output_c(3);
+ *      if (output_c.buffer_failure()) return -1; // Failure in buffer allocation
+ *
+ *      // Construction with a combination of a file descriptor and moving a
+ *      // pre-allocated unique_ptr of the buffer
+ *      dio::fileout output_d(3, std::move(buffer_2));
+ *
+ * Setting (or changing) the file descriptor after construction is available through
+ * the set_fd() function (See note in top of set_fd() func). opening and closing files
```

Comment:
Why is some documentation above the function, other documentation here?

It's fine to have general info here and more specific info above each function, but then it needs to be that way consistently, not just for some functions. Otherwise, put all documentation outside the class definition.

Reply:
(Note that in dinit generally the comments are attached to the function/class declaration, so this is really the preferable way. But it's not strictly necessary for this new library, I just want it to be consistent with itself at least).

Reply:
Documention about functions attached to functions except for constructor/destructor, inherited members, General error handling and exception-handling notes.